pandas/_libs/algos.pyx

```diff
@@ -793,216 +797,290 @@ ctypedef fused rank_t:
+            nan_fill_val = np.iinfo(np.int64).min
+        elif rank_t is uint64_t:
+            nan_fill_val = 0
+        else:
+            nan_fill_val = -np.inf
 
-    n = len(values)
-    ranks = np.empty(n, dtype='f8')
+        order = (masked_vals, ~mask, labels)
 
-    if rank_t is object:
-        _as = np.lexsort(keys=order)
-    else:
-        if tiebreak == TIEBREAK_FIRST:
-            # need to use a stable sort here
-            _as = np.lexsort(keys=order)
-            if not ascending:
-                tiebreak = TIEBREAK_FIRST_DESCENDING
-        else:
-            _as = np.lexsort(keys=order)
+    np.putmask(masked_vals, mask, nan_fill_val)
+
+    # lexsort using labels, then mask, then actual values
+    # each label corresponds to a different group value,
+    # the mask helps you differentiate missing values before
+    # performing sort on the actual values
+    _as = np.lexsort(order).astype(np.int64, copy=False)
 
     if not ascending:
         _as = _as[::-1]
 
-    sorted_data = values.take(_as)
-    sorted_mask = mask.take(_as)
-    _indices = np.diff(sorted_mask.astype(int)).nonzero()[0]
-    non_na_idx = _indices[0] if len(_indices) > 0 else -1
-    argsorted = _as.astype('i8')
-
+    # Loop over the length of the value array
+    # each incremental i value can be looked up in the _as array
+    # that we sorted previously, which gives us the location of
+    # that sorted value for retrieval back from the original
+    # values / masked_vals arrays
+    # TODO: de-duplicate once cython supports conditional nogil
     if rank_t is object:
-        # TODO: de-duplicate once cython supports conditional nogil
-        for i in range(n):
-            sum_ranks += i + 1
+        for i in range(N):
+            at_end = i == N - 1
+            # dups and sum_ranks will be incremented each loop where
+            # the value / group remains the same, and should be reset
+            # when either of those change
+            # Used to calculate tiebreakers
             dups += 1
-
-            val = sorted_data[i]
-
-            if rank_t is not uint64_t:
-                isnan = sorted_mask[i]
-                if isnan and keep_na:
-                    ranks[argsorted[i]] = NaN
-                    continue
-
-            count += 1.0
-
-            if rank_t is object:
-                condition = (
-                    i == n - 1 or
-                    are_diff(sorted_data[i + 1], val) or
-                    i == non_na_idx
-                )
+            sum_ranks += i - grp_start + 1
+
+            # Update out only when there is a transition of values or labels.
+            # When a new value or group is encountered, go back #dups steps(
+            # the number of occurrence of current value) and assign the ranks
+            # based on the starting index of the current group (grp_start)
+            # and the current index
+            if not at_end:
+                if rank_t is object:
```

Comment:
this check is redundant (check is redundant in master too)

Reply:
Yeah was wondering about that. Based on comments, seems like that was done to make future deduplication easier with conditional nogil. But that doesn't seem worth having unreachable/redundant code, removed those checks